File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        n = len(matrix)
        double = n * n
        half = double // 2
        sign, hmax = (-1, k) if k < half else (1, double - k + 1)
        h = []
        for row in matrix:
            for col in row:
                heappush(h, sign * col)
                if len(h) > hmax:
                    heappop(h)
        return sign * h[0]

File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        int squareN = n * n;
        
        PriorityQueue<Integer> heap;
        
        if(k > squareN / 2){
            heap = new PriorityQueue<>();
            k = squareN - k + 1;
        } else {
            heap = new PriorityQueue<>(
                (Integer o1, Integer o2) -> -Integer.compare(o1, o2)
            );
        }
        
        for(int i = 0; i < n; ++i){
            for(int j = 0; j < n; ++j){
                heap.add(matrix[i][j]);
                if(heap.size() > k){
                    heap.remove();
                }
            }
        }
        return heap.peek();
    }
}